### PR TITLE
Save work periodically to reduce loss from crashes

### LINKED ISSUE #929: Save work periodically to reduce loss from crashes
Try and find more times it would make sense to save, since it does likely take a little while to serialize and write the document which should ideally not cause annoying freezes for the user. We should aim to hopefully never make the user lose more than, say, 30 seconds of work.

Also, closing/reloading the page sometimes doesn't save correctly (maybe it doesn't have enough time if the editor is being slow or the document is big?). This should help in that case. Investigating the actual cause would also be nice.

Alternatively, now that we have moved toward the full node graph, it may be feasible to take a VS Code style approach where everything you do is saved in real time but the `*` isn't updated until you download the file. It might even be better to remove the `*` entirely, since whether you download the file isn't really that useful, except when #747 is in use (in a compatible browser and enabled by the user).

### PR DESCRIPTION
**Pull request by Comeza**
Closes #929

### CODE REVIEW

**Comment by Keavon** [comment]
Thanks for the PR! Please comment in the issue so GitHub lets me assign it to you. Looking forward to landing this!

**Comment by Keavon** [comment]
Thanks, I assigned it to you. Once you've finished implementing the feature and it seems to work in a production-ready way, please mark the PR as ready for review and also ping me so I notice and come to take a look with QA testing and code review. Welcome!

**Comment by TrueDoctor** [comment]
@Keavon are you d'acorde  with the approach taken here? Currently this saves the document after 30 s of inactivity. Should we add a second longer timer which forces a save if no inactivity was sufficiently long?

**Comment by Keavon** [comment]
I was thinking that maybe a good time to save would be, if 30 seconds have elapsed, when the history stack changes. Does that seem feasible?

**Comment by 0HyperCube** [comment]
!build

File: editor/src/messages/portfolio/portfolio_message_handler.rs [hunk 1]
```diff
@@ -76,8 +76,16 @@ impl MessageHandler<PortfolioMessage, (&InputPreprocessorMessageHandler, &Prefer
...
 					if let Some(document) = self.active_document_mut() {
 						document.set_auto_save_state(true);
+						responses.add(PortfolioMessage::AutoSaveDocument { document_id });
+					}
+				}
+			}
+			PortfolioMessage::AutoSaveAllDocuments => {
+				for (document_id, document) in self.documents.iter_mut() {
+					if !document.is_auto_saved() {
+						document.set_auto_save_state(true);
```
Comment (Keavon) [inline editor/src/messages/portfolio/portfolio_message_handler.rs:86]: Isn't this line redundant, since it happens in `AutoSaveDocument`?

Comment (Keavon) [reply editor/src/messages/portfolio/portfolio_message_handler.rs:86]: Disrgard.

**Review by Keavon — APPROVED** [review]
Sorry this took so abnormally long to review!

**Pull request merged**